diff --git a/yazi-plugin/src/elements/style.rs b/yazi-plugin/src/elements/style.rs
index 3bf3b7b..f197f35 100644
--- a/yazi-plugin/src/elements/style.rs
+++ b/yazi-plugin/src/elements/style.rs
@@ -76,6 +76,10 @@ impl UserData for Style {
 			ud.borrow_mut::<Self>()?.0.add_modifier |= ratatui::style::Modifier::CROSSED_OUT;
 			Ok(ud)
 		});
+		methods.add_function("reverse", |_, ud: AnyUserData| {
+			ud.borrow_mut::<Self>()?.0.add_modifier |= ratatui::style::Modifier::REVERSED;
+			Ok(ud)
+		});
 		methods.add_function("reset", |_, ud: AnyUserData| {
 			ud.borrow_mut::<Self>()?.0.add_modifier = ratatui::style::Modifier::empty();
 			Ok(ud)
